In [1]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from typing import List, Dict, Tuple, Any

In [2]:
from sklearn.metrics import classification_report, accuracy_score, recall_score, f1_score, precision_score

In [3]:
X_train = pd.read_parquet('train_binarized.parquet')
X_val = pd.read_parquet('val_binarized.parquet')

In [4]:
y_train = pd.read_parquet('y_train.parquet')
y_val = pd.read_parquet('y_val.parquet')

In [5]:
X_train

pregnancies_ge_1  pregnancies_le_1  pregnancies_ge_3  pregnancies_le_3  \
357                 1                 0                 1                 0   
73                  1                 0                 1                 0   
352                 1                 0                 1                 1   
497                 1                 0                 0                 1   
145                 0                 1                 0                 1   
..                ...               ...               ...               ...   
71                  1                 0                 1                 0   
106                 1                 1                 0                 1   
270                 1                 0                 1                 0   
435                 0                 1                 0                 1   
102                 0                 1                 0                 1   

     pregnancies_ge_7  pregnancies_le_7  pregnancies_ge_9  pregnancies_le_9  \
357                 1                 0                 1                 0   
73                  0                 1                 0                 1   
352                 0                 1                 0                 1   
497                 0                 1                 0                 1   
145                 0                 1                 0                 1   
..                ...               ...               ...               ...   
71                  0                 1                 0                 1   
106                 0                 1                 0                 1   
270                 1                 0                 1                 0   
435                 0                 1                 0                 1   
102                 0                 1                 0                 1   

     glucose_ge_79  glucose_le_79  ...  age_ge_24  age_le_24  age_ge_29  \
357              1              0  ...          1          0          1   
73               1              0  ...          0          1          0   
352              0              1  ...          1          0          1   
497              1              0  ...          1          0          0   
145              1              0  ...          0          1          0   
..             ...            ...  ...        ...        ...        ...   
71               1              0  ...          1          0          0   
106              1              0  ...          1          0          0   
270              1              0  ...          1          0          1   
435              1              0  ...          1          0          1   
102              1              0  ...          0          1          0   

     age_le_29  age_ge_40  age_le_40  age_ge_50  age_le_50  age_ge_60  \
357          0          1          0          0          1          0   
73           1          0          1          0          1          0   
352          0          1          0          0          1          0   
497          1          0          1          0          1          0   
145          1          0          1          0          1          0   
..         ...        ...        ...        ...        ...        ...   
71           1          0          1          0          1          0   
106          1          0          1          0          1          0   
270          0          0          1          0          1          0   
435          1          0          1          0          1          0   
102          1          0          1          0          1          0   

     age_le_60  
357          1  
73           1  
352          1  
497          1  
145          1  
..         ...  
71           1  
106          1  
270          1  
435          1  
102          1  

[576 rows x 78 columns]

In [6]:
y_train

outcome
357        1
73         0
352        0
497        0
145        0
..       ...
71         0
106        0
270        1
435        1
102        0

[576 rows x 1 columns]

In [114]:
X_val

pregnancies_ge_1  pregnancies_le_1  pregnancies_ge_3  pregnancies_le_3  \
668                 1                 0                 1                 0   
324                 1                 0                 0                 1   
624                 1                 0                 0                 1   
690                 1                 0                 1                 0   
473                 1                 0                 1                 0   
..                ...               ...               ...               ...   
554                 1                 1                 0                 1   
319                 1                 0                 1                 0   
594                 1                 0                 1                 0   
6                   1                 0                 1                 1   
615                 1                 0                 1                 1   

     pregnancies_ge_7  pregnancies_le_7  pregnancies_ge_9  pregnancies_le_9  \
668                 0                 1                 0                 1   
324                 0                 1                 0                 1   
624                 0                 1                 0                 1   
690                 1                 0                 0                 1   
473                 1                 1                 0                 1   
..                ...               ...               ...               ...   
554                 0                 1                 0                 1   
319                 0                 1                 0                 1   
594                 0                 1                 0                 1   
6                   0                 1                 0                 1   
615                 0                 1                 0                 1   

     glucose_ge_79  glucose_le_79  ...  age_ge_24  age_le_24  age_ge_29  \
668              1              0  ...          1          0          1   
324              1              0  ...          0          1          0   
624              1              0  ...          0          1          0   
690              1              0  ...          1          0          1   
473              1              0  ...          1          0          1   
..             ...            ...  ...        ...        ...        ...   
554              1              0  ...          1          0          0   
319              1              0  ...          1          0          1   
594              1              0  ...          1          0          1   
6                0              1  ...          1          0          0   
615              1              0  ...          1          0          0   

     age_le_29  age_ge_40  age_le_40  age_ge_50  age_le_50  age_ge_60  \
668          0          1          0          0          1          0   
324          1          0          1          0          1          0   
624          1          0          1          0          1          0   
690          0          0          1          0          1          0   
473          0          1          0          1          1          0   
..         ...        ...        ...        ...        ...        ...   
554          1          0          1          0          1          0   
319          0          1          0          1          0          0   
594          0          0          1          0          1          0   
6            1          0          1          0          1          0   
615          1          0          1          0          1          0   

     age_le_60  
668          1  
324          1  
624          1  
690          1  
473          1  
..         ...  
554          1  
319          1  
594          1  
6            1  
615          1  

[192 rows x 78 columns]

In [8]:
X_train_pos = X_train[y_train['outcome'] == 1]
X_train_neg = X_train[y_train['outcome'] == 0]

In [9]:
X_train_pos.shape, X_train_neg.shape

((199, 78), (377, 78))

In [10]:
y_val_pred = pd.DataFrame(np.zeros_like(y_val), columns=['prediction'], index=y_val.index)
y_val_pred

prediction
668           0
324           0
624           0
690           0
473           0
..          ...
554           0
319           0
594           0
6             0
615           0

[192 rows x 1 columns]

In [11]:
for x_test in X_val.iterrows():
    pos_imp = 0
    for x_pos in X_train_pos.iterrows():
        pos_imp += (x_test[1] * x_pos[1]).sum()
    
    pos_imp /= X_train_pos.shape[0]

    neg_imp = 0
    for x_neg in X_train_neg.iterrows():
        neg_imp += (x_test[1] * x_neg[1]).sum()
    
    neg_imp /= X_train_neg.shape[0]

    y_val_pred.loc[x_test[0], 'prediction'] = (pos_imp - neg_imp > 0).astype(float)

In [12]:
y_val_pred

prediction
668           1
324           0
624           0
690           0
473           1
..          ...
554           0
319           1
594           1
6             0
615           0

[192 rows x 1 columns]

In [13]:
y_val

outcome
668        0
324        0
624        0
690        0
473        0
..       ...
554        0
319        1
594        0
6          1
615        0

[192 rows x 1 columns]

In [14]:
print(classification_report(y_val, y_val_pred))

              precision    recall  f1-score   support

           0       0.81      0.74      0.77       123
           1       0.59      0.68      0.64        69

    accuracy                           0.72       192
   macro avg       0.70      0.71      0.70       192
weighted avg       0.73      0.72      0.72       192



In [15]:
f1_score(y_val, y_val_pred)

0.6351351351351351

In [16]:
recall_score(y_val, y_val_pred)

0.6811594202898551

In [17]:
precision_score(y_val, y_val_pred)

0.5949367088607594

In [18]:
accuracy_score(y_val, y_val_pred)

0.71875

In [19]:
from sklearn.base import BaseEstimator, ClassifierMixin
from sklearn.utils.validation import check_X_y, check_array, check_is_fitted

In [20]:
class LazyFCA(ClassifierMixin, BaseEstimator):
    def __init__(self):
        pass

    def fit(self, X, y):
        self.X_train = X
        self.y_train = y 

        self.X_train_pos = X[y['outcome'] == 1]
        self.X_train_neg = X[y['outcome'] == 0]

        return self

    def predict(self, X):
        y_val_pred = pd.DataFrame(np.zeros(X.shape[0]), columns=['prediction'], index=X.index)

        for x_test in X.iterrows():
            pos_imp = 0
            for x_pos in self.X_train_pos.iterrows():
                pos_imp += (x_test[1] * x_pos[1]).sum()
            
            pos_imp /= self.X_train_pos.shape[0]

            neg_imp = 0
            for x_neg in self.X_train_neg.iterrows():
                neg_imp += (x_test[1] * x_neg[1]).sum()
            
            neg_imp /= self.X_train_neg.shape[0]

            y_val_pred.loc[x_test[0], 'prediction'] = (pos_imp - neg_imp > 0).astype(float)
        
        return y_val_pred

In [21]:
lazy_clf = LazyFCA()

In [22]:
lazy_clf.fit(X_train, y_train)

LazyFCA()

In [23]:
y_pred = lazy_clf.predict(X_val)

In [24]:
print(classification_report(y_val, y_pred))

              precision    recall  f1-score   support

           0       0.81      0.74      0.77       123
           1       0.59      0.68      0.64        69

    accuracy                           0.72       192
   macro avg       0.70      0.71      0.70       192
weighted avg       0.73      0.72      0.72       192



## Code from Alan

In [25]:
class LazyClassifierFCA:
    def __init__(self):
        self.X_train = None
        self.y_train = None

    def fit(self, X_train: pd.DataFrame, y_train: pd.Series) -> None:
        self.X_train = X_train
        self.y_train = y_train
        
    def classify_sample(self, sample: pd.Series) -> Any:
        # Split X_train into positive and negative classes
        X_train_positive = self.X_train[y_train['outcome'] == 1]
        X_train_negative = self.X_train[y_train['outcome'] == 0]
        
        positive_classifiers = 0
        negative_classifiers = 0
        
        # Function to check if intersection with a train sample is a positive classifier
        def is_positive_classifier(intersection):
            # Find samples in X_train_positive that contain the intersection
            num_positive = ((X_train_positive | (~intersection.astype(bool)).astype(np.int32)) == True).all(axis=1).sum()
            num_negative = ((X_train_negative | (~intersection.astype(bool)).astype(np.int32)) == True).all(axis=1).sum()
            return num_negative == 0 and num_positive > 1
        
        # Function to check if intersection is a negative classifier
        def is_negative_classifier(intersection):
            # Find samples in X_train_negative that contain the intersection
            num_positive = ((X_train_positive | (~intersection.astype(bool)).astype(np.int32)) == True).all(axis=1).sum()
            num_negative = ((X_train_negative | (~intersection.astype(bool)).astype(np.int32)) == True).all(axis=1).sum()
            return num_positive == 0 and num_negative > 1
        
        # Check for positive classifiers by intersecting sample with each positive object
        for _, pos_sample in X_train_positive.iterrows():
            intersection = sample & pos_sample
            if is_positive_classifier(intersection):
                positive_classifiers += 1

        # Check for negative classifiers by intersecting sample with each negative object
        for _, neg_sample in X_train_negative.iterrows():
            intersection = sample & neg_sample
            if is_negative_classifier(intersection):
                negative_classifiers += 1

        # Determine the class based on the number of classifiers
        if positive_classifiers > negative_classifiers:
            print(f"sample {sample.name} is classified as 1, {positive_classifiers=}, {negative_classifiers=}")
            return 1  # Predict positive
            
        elif negative_classifiers > positive_classifiers:
            print(f"sample {sample.name} is classified as 0, {positive_classifiers=}, {negative_classifiers=}")
            return 0  # Predict negative

        else:
            # If equal, you can decide on a rule, like defaulting to 0 or 1, or returning 'undetermined'
            print(f"sample {sample.name} is classified as 1, default, {positive_classifiers=}, {negative_classifiers=}")
        
            return 1  # or 0, depending on the choice


    def predict(self, X_test: pd.DataFrame) -> List[Any]:
        # List to store predictions for each test sample
        predictions = []
        
        # Iterate through each sample in X_test
        for _, sample in X_test.iterrows():
            # Classify the sample and append the result to predictions
            prediction = self.classify_sample(sample)
            predictions.append(prediction)
        
        return predictions

In [26]:
X_train_positive = X_train[y_train['outcome'] == 1]
X_train_negative = X_train[y_train['outcome'] == 0]

In [27]:
X_train_positive

pregnancies_ge_1  pregnancies_le_1  pregnancies_ge_3  pregnancies_le_3  \
357                 1                 0                 1                 0   
291                 0                 1                 0                 1   
132                 1                 0                 1                 1   
719                 1                 0                 1                 0   
56                  1                 0                 1                 0   
..                ...               ...               ...               ...   
458                 1                 0                 1                 0   
214                 1                 0                 1                 0   
614                 1                 0                 1                 0   
270                 1                 0                 1                 0   
435                 0                 1                 0                 1   

     pregnancies_ge_7  pregnancies_le_7  pregnancies_ge_9  pregnancies_le_9  \
357                 1                 0                 1                 0   
291                 0                 1                 0                 1   
132                 0                 1                 0                 1   
719                 0                 1                 0                 1   
56                  1                 1                 0                 1   
..                ...               ...               ...               ...   
458                 1                 0                 1                 0   
214                 1                 0                 1                 1   
614                 1                 0                 1                 0   
270                 1                 0                 1                 0   
435                 0                 1                 0                 1   

     glucose_ge_79  glucose_le_79  ...  age_ge_24  age_le_24  age_ge_29  \
357              1              0  ...          1          0          1   
291              1              0  ...          1          0          0   
132              1              0  ...          1          0          1   
719              1              0  ...          1          0          1   
56               1              0  ...          1          0          1   
..             ...            ...  ...        ...        ...        ...   
458              1              0  ...          1          0          1   
214              1              0  ...          1          0          1   
614              1              0  ...          1          0          1   
270              1              0  ...          1          0          1   
435              1              0  ...          1          0          1   

     age_le_29  age_ge_40  age_le_40  age_ge_50  age_le_50  age_ge_60  \
357          0          1          0          0          1          0   
291          1          0          1          0          1          0   
132          0          0          1          0          1          0   
719          0          1          0          1          0          0   
56           0          1          0          0          1          0   
..         ...        ...        ...        ...        ...        ...   
458          0          1          0          1          0          0   
214          0          0          1          0          1          0   
614          0          1          0          1          1          0   
270          0          0          1          0          1          0   
435          1          0          1          0          1          0   

     age_le_60  
357          1  
291          1  
132          1  
719          1  
56           1  
..         ...  
458          1  
214          1  
614          1  
270          1  
435          1  

[199 rows x 78 columns]

In [28]:
positive_classifiers = 0
negative_classifiers = 0

In [29]:
def is_positive_classifier(intersection):
    # Find samples in X_train_positive that contain the intersection
    num_positive = ((X_train_positive | ~intersection) == True).all(axis=1).sum()
    num_negative = ((X_train_negative | ~intersection) == True).all(axis=1).sum()
    return num_negative == 0 and num_positive > 1

# Function to check if intersection is a negative classifier
def is_negative_classifier(intersection):
    # Find samples in X_train_negative that contain the intersection
    num_positive = ((X_train_positive | ~intersection) == True).all(axis=1).sum()
    num_negative = ((X_train_negative | ~intersection) == True).all(axis=1).sum()
    return num_positive == 0 and num_negative > 1

In [30]:
for _, pos_sample in X_train_positive.iterrows():
    break

In [31]:
pos_sample

pregnancies_ge_1    1
pregnancies_le_1    0
pregnancies_ge_3    1
pregnancies_le_3    0
pregnancies_ge_7    1
                   ..
age_le_40           0
age_ge_50           0
age_le_50           1
age_ge_60           0
age_le_60           1
Name: 357, Length: 78, dtype: int32

In [32]:
sample = X_val.iloc[1]
sample

pregnancies_ge_1    1
pregnancies_le_1    0
pregnancies_ge_3    0
pregnancies_le_3    1
pregnancies_ge_7    0
                   ..
age_le_40           1
age_ge_50           0
age_le_50           1
age_ge_60           0
age_le_60           1
Name: 324, Length: 78, dtype: int32

In [33]:
intersection = sample & pos_sample

In [34]:
sample.sum(), pos_sample.sum()

(40, 39)

In [35]:
(~intersection.astype(bool)).astype(np.int32)

pregnancies_ge_1    0
pregnancies_le_1    1
pregnancies_ge_3    1
pregnancies_le_3    1
pregnancies_ge_7    1
                   ..
age_le_40           1
age_ge_50           1
age_le_50           0
age_ge_60           1
age_le_60           0
Length: 78, dtype: int32

In [36]:
intersection.sum()

25

In [37]:
num_positive = ((X_train_positive |(~intersection.astype(bool)).astype(np.int32)) == True).all(axis=1).sum()

In [38]:
num_positive

3

In [39]:
((X_train_positive | ~intersection) == True).all(axis=1).sum()

0

In [40]:
X_train_positive

pregnancies_ge_1  pregnancies_le_1  pregnancies_ge_3  pregnancies_le_3  \
357                 1                 0                 1                 0   
291                 0                 1                 0                 1   
132                 1                 0                 1                 1   
719                 1                 0                 1                 0   
56                  1                 0                 1                 0   
..                ...               ...               ...               ...   
458                 1                 0                 1                 0   
214                 1                 0                 1                 0   
614                 1                 0                 1                 0   
270                 1                 0                 1                 0   
435                 0                 1                 0                 1   

     pregnancies_ge_7  pregnancies_le_7  pregnancies_ge_9  pregnancies_le_9  \
357                 1                 0                 1                 0   
291                 0                 1                 0                 1   
132                 0                 1                 0                 1   
719                 0                 1                 0                 1   
56                  1                 1                 0                 1   
..                ...               ...               ...               ...   
458                 1                 0                 1                 0   
214                 1                 0                 1                 1   
614                 1                 0                 1                 0   
270                 1                 0                 1                 0   
435                 0                 1                 0                 1   

     glucose_ge_79  glucose_le_79  ...  age_ge_24  age_le_24  age_ge_29  \
357              1              0  ...          1          0          1   
291              1              0  ...          1          0          0   
132              1              0  ...          1          0          1   
719              1              0  ...          1          0          1   
56               1              0  ...          1          0          1   
..             ...            ...  ...        ...        ...        ...   
458              1              0  ...          1          0          1   
214              1              0  ...          1          0          1   
614              1              0  ...          1          0          1   
270              1              0  ...          1          0          1   
435              1              0  ...          1          0          1   

     age_le_29  age_ge_40  age_le_40  age_ge_50  age_le_50  age_ge_60  \
357          0          1          0          0          1          0   
291          1          0          1          0          1          0   
132          0          0          1          0          1          0   
719          0          1          0          1          0          0   
56           0          1          0          0          1          0   
..         ...        ...        ...        ...        ...        ...   
458          0          1          0          1          0          0   
214          0          0          1          0          1          0   
614          0          1          0          1          1          0   
270          0          0          1          0          1          0   
435          1          0          1          0          1          0   

     age_le_60  
357          1  
291          1  
132          1  
719          1  
56           1  
..         ...  
458          1  
214          1  
614          1  
270          1  
435          1  

[199 rows x 78 columns]

In [41]:
classifier = LazyClassifierFCA()
classifier.fit(X_train, y_train)

In [42]:
y_train['outcome']

357    1
73     0
352    0
497    0
145    0
      ..
71     0
106    0
270    1
435    1
102    0
Name: outcome, Length: 576, dtype: int64

In [43]:
sample = X_val.iloc[1]
sample

pregnancies_ge_1    1
pregnancies_le_1    0
pregnancies_ge_3    0
pregnancies_le_3    1
pregnancies_ge_7    0
                   ..
age_le_40           1
age_ge_50           0
age_le_50           1
age_ge_60           0
age_le_60           1
Name: 324, Length: 78, dtype: int32

In [44]:
classifier.classify_sample(sample)

sample 324 is classified as 0, positive_classifiers=16, negative_classifiers=135


0

In [45]:
y_pred = classifier.predict(X_val)

sample 668 is classified as 0, positive_classifiers=26, negative_classifiers=130
sample 324 is classified as 0, positive_classifiers=16, negative_classifiers=135
sample 624 is classified as 0, positive_classifiers=1, negative_classifiers=190
sample 690 is classified as 0, positive_classifiers=28, negative_classifiers=87
sample 473 is classified as 1, positive_classifiers=33, negative_classifiers=30
sample 204 is classified as 0, positive_classifiers=26, negative_classifiers=82
sample 97 is classified as 0, positive_classifiers=0, negative_classifiers=278
sample 336 is classified as 0, positive_classifiers=46, negative_classifiers=93
sample 568 is classified as 0, positive_classifiers=42, negative_classifiers=69
sample 148 is classified as 1, positive_classifiers=74, negative_classifiers=2
sample 667 is classified as 0, positive_classifiers=18, negative_classifiers=112
sample 212 is classified as 1, positive_classifiers=43, negative_classifiers=24
sample 199 is classified as 0, positive

In [46]:
print(classification_report(y_val, y_pred))

              precision    recall  f1-score   support

           0       0.72      0.91      0.81       123
           1       0.70      0.38      0.49        69

    accuracy                           0.72       192
   macro avg       0.71      0.64      0.65       192
weighted avg       0.72      0.72      0.69       192



In [69]:
class LazyClassifierFCA2:
    def __init__(self, max_counter_examples=5):
        self.X_train = None
        self.y_train = None
        self.max_counter_examples = max_counter_examples

    def fit(self, X_train: pd.DataFrame, y_train: pd.Series) -> None:
        self.X_train = X_train
        self.y_train = y_train
        
    def classify_sample(self, sample: pd.Series) -> Any:
        # Split X_train into positive and negative classes
        X_train_positive = self.X_train[y_train['outcome'] == 1]
        X_train_negative = self.X_train[y_train['outcome'] == 0]
        
        positive_classifiers = 0
        negative_classifiers = 0
        
        # Function to check if intersection with a train sample is a positive classifier
        def is_positive_classifier(intersection):
            # Find samples in X_train_positive that contain the intersection
            num_positive = ((X_train_positive | (~intersection.astype(bool)).astype(np.int32)) == True).all(axis=1).sum()
            num_negative = ((X_train_negative | (~intersection.astype(bool)).astype(np.int32)) == True).all(axis=1).sum()
            return num_negative < self.max_counter_examples and num_positive > 1
        
        # Function to check if intersection is a negative classifier
        def is_negative_classifier(intersection):
            # Find samples in X_train_negative that contain the intersection
            num_positive = ((X_train_positive | (~intersection.astype(bool)).astype(np.int32)) == True).all(axis=1).sum()
            num_negative = ((X_train_negative | (~intersection.astype(bool)).astype(np.int32)) == True).all(axis=1).sum()
            return num_positive < self.max_counter_examples and num_negative > 1
        
        # Check for positive classifiers by intersecting sample with each positive object
        for _, pos_sample in X_train_positive.iterrows():
            intersection = sample & pos_sample
            if is_positive_classifier(intersection):
                positive_classifiers += 1

        # Check for negative classifiers by intersecting sample with each negative object
        for _, neg_sample in X_train_negative.iterrows():
            intersection = sample & neg_sample
            if is_negative_classifier(intersection):
                negative_classifiers += 1

        # Determine the class based on the number of classifiers
        if positive_classifiers / X_train_positive.shape[0] > negative_classifiers / X_train_negative.shape[0]:
            print(f"sample {sample.name} is classified as 1, {positive_classifiers=}, {negative_classifiers=}")
            return 1  # Predict positive
            
        elif negative_classifiers / X_train_negative.shape[0] > positive_classifiers / X_train_positive.shape[0]:
            print(f"sample {sample.name} is classified as 0, {positive_classifiers=}, {negative_classifiers=}")
            return 0  # Predict negative

        else:
            # If equal, you can decide on a rule, like defaulting to 0 or 1, or returning 'undetermined'
            print(f"sample {sample.name} is classified as 1, default, {positive_classifiers=}, {negative_classifiers=}")
        
            return 1  # or 0, depending on the choice


    def predict(self, X_test: pd.DataFrame) -> List[Any]:
        # List to store predictions for each test sample
        predictions = []
        
        # Iterate through each sample in X_test
        for _, sample in X_test.iterrows():
            # Classify the sample and append the result to predictions
            prediction = self.classify_sample(sample)
            predictions.append(prediction)
        
        return predictions

In [70]:
pd.value_counts(y_train['outcome'])

C:\Users\yunes\AppData\Local\Temp\ipykernel_15764\263467042.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  pd.value_counts(y_train['outcome'])


outcome
0    377
1    199
Name: count, dtype: int64

In [77]:
classifier2 = LazyClassifierFCA2(max_counter_examples=3)
classifier2.fit(X_train, y_train)

In [78]:
y_pred2 = classifier2.predict(X_val)

sample 668 is classified as 0, positive_classifiers=53, negative_classifiers=199
sample 324 is classified as 0, positive_classifiers=37, negative_classifiers=229
sample 624 is classified as 0, positive_classifiers=20, negative_classifiers=271
sample 690 is classified as 0, positive_classifiers=71, negative_classifiers=185
sample 473 is classified as 1, positive_classifiers=74, negative_classifiers=100
sample 204 is classified as 0, positive_classifiers=58, negative_classifiers=177
sample 97 is classified as 0, positive_classifiers=2, negative_classifiers=321
sample 336 is classified as 1, positive_classifiers=103, negative_classifiers=162
sample 568 is classified as 0, positive_classifiers=84, negative_classifiers=208
sample 148 is classified as 1, positive_classifiers=136, negative_classifiers=47
sample 667 is classified as 0, positive_classifiers=64, negative_classifiers=223
sample 212 is classified as 1, positive_classifiers=75, negative_classifiers=99
sample 199 is classified as 0,

In [79]:
print(classification_report(y_val, y_pred2))

              precision    recall  f1-score   support

           0       0.72      0.87      0.79       123
           1       0.63      0.39      0.48        69

    accuracy                           0.70       192
   macro avg       0.67      0.63      0.63       192
weighted avg       0.69      0.70      0.68       192



Results of max_counter_examples=3 and division by number of objects in class:
precision    recall  f1-score   support

           0       0.72      0.87      0.79       123
           1       0.63      0.39      0.48        69

    accuracy                           0.70       192
   macro avg       0.67      0.63      0.63       192
weighted avg       0.69      0.70      0.68       192


Results of max_counter_examples=2 and division by number of objects in class:
precision    recall  f1-score   support

           0       0.75      0.86      0.80       123
           1       0.66      0.48      0.55        69

    accuracy                           0.72       192
   macro avg       0.70      0.67      0.68       192
weighted avg       0.72      0.72      0.71       192

Results of max_counter_examples=1 and division by number of objects in class:
 precision    recall  f1-score   support

           0       0.78      0.84      0.81       123
           1       0.67      0.58      0.62        69

    accuracy                           0.74       192
   macro avg       0.72      0.71      0.71       192
weighted avg       0.74      0.74      0.74       192

Results of max_counter_examples=1
precision    recall  f1-score   support

           0       0.72      0.91      0.81       123
           1       0.70      0.38      0.49        69

    accuracy                           0.72       192
   macro avg       0.71      0.64      0.65       192
weighted avg       0.72      0.72      0.69       192

Results of max_counter_examples=2
precision    recall  f1-score   support

           0       0.68      0.93      0.78       123
           1       0.64      0.20      0.31        69

    accuracy                           0.67       192
   macro avg       0.66      0.57      0.55       192
weighted avg       0.66      0.67      0.61       192

Results of max_counter_examples=3
precision    recall  f1-score   support

           0       0.66      0.96      0.78       123
           1       0.58      0.10      0.17        69

    accuracy                           0.65       192
   macro avg       0.62      0.53      0.48       192
weighted avg       0.63      0.65      0.56       192

Results of max_counter_examples=4
 precision    recall  f1-score   support

           0       0.65      0.97      0.78       123
           1       0.50      0.06      0.10        69

    accuracy                           0.64       192
   macro avg       0.57      0.51      0.44       192
weighted avg       0.59      0.64      0.53       192

Results of max_counter_examples=5 
 precision    recall  f1-score   support

           0       0.65      0.98      0.78       123
           1       0.60      0.04      0.08        69

    accuracy                           0.65       192
   macro avg       0.62      0.51      0.43       192
weighted avg       0.63      0.65      0.53       192


Results of max_counter_examples=8
precision    recall  f1-score   support

           0       0.64      1.00      0.78       123
           1       0.00      0.00      0.00        69

    accuracy                           0.64       192
   macro avg       0.32      0.50      0.39       192
weighted avg       0.41      0.64      0.50       192

In [97]:
class LazyClassifierFCA3:
    def __init__(self, max_counter_examples=5, min_cardinality=25):
        self.X_train = None
        self.y_train = None
        self.max_counter_examples = max_counter_examples
        self.min_cardinality = min_cardinality

    def fit(self, X_train: pd.DataFrame, y_train: pd.Series) -> None:
        self.X_train = X_train
        self.y_train = y_train
        
    def classify_sample(self, sample: pd.Series) -> Any:
        # Split X_train into positive and negative classes
        X_train_positive = self.X_train[y_train['outcome'] == 1]
        X_train_negative = self.X_train[y_train['outcome'] == 0]
        
        positive_classifiers = 0
        negative_classifiers = 0
        
        # Function to check if intersection with a train sample is a positive classifier
        def is_positive_classifier(intersection):
            # Find samples in X_train_positive that contain the intersection
            num_positive = ((X_train_positive | (~intersection.astype(bool)).astype(np.int32)) == True).all(axis=1).sum()
            num_negative = ((X_train_negative | (~intersection.astype(bool)).astype(np.int32)) == True).all(axis=1).sum()
            return num_negative < self.max_counter_examples and num_positive > 1
        
        # Function to check if intersection is a negative classifier
        def is_negative_classifier(intersection):
            # Find samples in X_train_negative that contain the intersection
            num_positive = ((X_train_positive | (~intersection.astype(bool)).astype(np.int32)) == True).all(axis=1).sum()
            num_negative = ((X_train_negative | (~intersection.astype(bool)).astype(np.int32)) == True).all(axis=1).sum()
            return num_positive < self.max_counter_examples and num_negative > 1
        
        # Check for positive classifiers by intersecting sample with each positive object
        for _, pos_sample in X_train_positive.iterrows():
            intersection = sample & pos_sample
            if is_positive_classifier(intersection) and intersection.sum() >= self.min_cardinality:
                positive_classifiers += 1

        # Check for negative classifiers by intersecting sample with each negative object
        for _, neg_sample in X_train_negative.iterrows():
            intersection = sample & neg_sample
            if is_negative_classifier(intersection) and intersection.sum() >= self.min_cardinality:
                negative_classifiers += 1

        # Determine the class based on the number of classifiers
        if positive_classifiers / X_train_positive.shape[0] > negative_classifiers / X_train_negative.shape[0]:
            print(f"sample {sample.name} is classified as 1, {positive_classifiers=}, {negative_classifiers=}")
            return 1  # Predict positive
            
        elif negative_classifiers / X_train_negative.shape[0] > positive_classifiers / X_train_positive.shape[0]:
            print(f"sample {sample.name} is classified as 0, {positive_classifiers=}, {negative_classifiers=}")
            return 0  # Predict negative

        else:
            # If equal, you can decide on a rule, like defaulting to 0 or 1, or returning 'undetermined'
            print(f"sample {sample.name} is classified as 1, default, {positive_classifiers=}, {negative_classifiers=}")
        
            return 1  # or 0, depending on the choice


    def predict(self, X_test: pd.DataFrame) -> List[Any]:
        # List to store predictions for each test sample
        predictions = []
        
        # Iterate through each sample in X_test
        for _, sample in X_test.iterrows():
            # Classify the sample and append the result to predictions
            prediction = self.classify_sample(sample)
            predictions.append(prediction)
        
        return predictions

In [98]:
X_train.shape

(576, 78)

In [111]:
classifier3 = LazyClassifierFCA3(max_counter_examples=1, min_cardinality=25)
classifier3.fit(X_train, y_train)

In [112]:
y_pred3 = classifier3.predict(X_val)

sample 668 is classified as 0, positive_classifiers=20, negative_classifiers=108
sample 324 is classified as 0, positive_classifiers=15, negative_classifiers=131
sample 624 is classified as 0, positive_classifiers=1, negative_classifiers=185
sample 690 is classified as 0, positive_classifiers=26, negative_classifiers=83
sample 473 is classified as 1, positive_classifiers=28, negative_classifiers=25
sample 204 is classified as 0, positive_classifiers=23, negative_classifiers=77
sample 97 is classified as 0, positive_classifiers=0, negative_classifiers=258
sample 336 is classified as 0, positive_classifiers=27, negative_classifiers=83
sample 568 is classified as 1, positive_classifiers=40, negative_classifiers=69
sample 148 is classified as 1, positive_classifiers=48, negative_classifiers=2
sample 667 is classified as 0, positive_classifiers=14, negative_classifiers=103
sample 212 is classified as 1, positive_classifiers=31, negative_classifiers=20
sample 199 is classified as 0, positive

In [113]:
print(classification_report(y_val, y_pred3))

              precision    recall  f1-score   support

           0       0.75      0.82      0.79       123
           1       0.62      0.52      0.57        69

    accuracy                           0.71       192
   macro avg       0.69      0.67      0.68       192
weighted avg       0.71      0.71      0.71       192



Results with min_cardinality = 25, max_counter_examples = 1:

precision    recall  f1-score   support

           0       0.75      0.82      0.79       123
           1       0.62      0.52      0.57        69

    accuracy                           0.71       192
   macro avg       0.69      0.67      0.68       192
weighted avg       0.71      0.71      0.71       192

Results with min_cardinality = 35, max_counter_examples = 1:

precision    recall  f1-score   support

           0       0.83      0.51      0.63       123
           1       0.48      0.81      0.61        69

    accuracy                           0.62       192
   macro avg       0.66      0.66      0.62       192
weighted avg       0.70      0.62      0.62       192

Results with min_cardinality = 30, max_counter_examples = 1:
 precision    recall  f1-score   support

           0       0.77      0.82      0.80       123
           1       0.64      0.57      0.60        69

    accuracy                           0.73       192
   macro avg       0.71      0.69      0.70       192
weighted avg       0.72      0.73      0.73       192

Results with min_cardinality = 20, max_counter_examples = 1:
precision    recall  f1-score   support

           0       0.78      0.84      0.81       123
           1       0.67      0.58      0.62        69

    accuracy                           0.74       192
   macro avg       0.72      0.71      0.71       192
weighted avg       0.74      0.74      0.74       192

Results with min_cardinality = 25, max_counter_examples = 1:
 precision    recall  f1-score   support

           0       0.75      0.82      0.79       123
           1       0.62      0.52      0.57        69

    accuracy                           0.71       192
   macro avg       0.69      0.67      0.68       192
weighted avg       0.71      0.71      0.71       192